### ISSUE TITLE
TpFIBQuery. закрытие транзакции после первого фетча

### ISSUE THREAD

**Issue by dshumko** [madorin/fibplus#4]
если у TpFIBQuery задать следующие параметры 
GoToFirstRecordOnExecute = True;
qoStartTransaction = true
qoAutoCommit = true
то при выборке `select` происходит закрытие транзакции после первого фетча, что приводит к ошибке выполнения

смотрим код в версия [7.6 код](https://github.com/madorin/fibplus/blob/master/FIBQuery.pas#L4684-L4689)
```pascal
if FGoToFirstRecordOnExecute
then begin
    DoAfterExecute(Assigned(FMDTMainDataOrder), vAfterExecEvent);
    vAfterExecEvent := False;
    Next;
end;
```
в версии 7.5 код был такой
```pascal
if FGoToFirstRecordOnExecute then
    Next;
```
думаю `DoAfterExecute` лишний в версии 7.6 этот код выполнится [позже](https://github.com/madorin/fibplus/blob/master/FIBQuery.pas#L4742-L4743)

**Comment by madorin** [comment]
Да, там проблема, при select стэйтментов он все равно делает commit в DoAfterExecute.
Позже этот код не выполняется так как ставится флаг vAfterExecEvent.
Буду переделывать как у датасетов: Если select то не сделать Commit.
Наверное это лучший компроммис когда AutoCommit.

**Comment by vkrapotkin** [comment]
я правильно понимаю, что тут может помочь пока только использование FetchAll ?

**Comment by dshumko** [comment]
FetchAll нет в FIBQuery.
для себя закомментировал  
```pascal
if FGoToFirstRecordOnExecute
then begin
    // DoAfterExecute(Assigned(FMDTMainDataOrder), vAfterExecEvent);
    // vAfterExecEvent := False;
    Next;
end;
```

**Comment by vkrapotkin** [comment]
точно. посмотрел. это в Dataset
реализовано как Bookmark + Last

**Comment by AnryJester** [comment]
Данный фикс "сломал" функционал получения количества записей, который был раньше работал корректно и вполне логично:
В событии OnAfterExecute проверяется количество записей через свойство RecordCount.
Соответственно перенос кода
```
if FGoToFirstRecordOnExecute and FOpen then
    Next;
```
после вызова события DoAfterExecute не логично и "ломает" работу компонента.
1. При проверке количества записей в событии DoAfterExecute соответственно получаем 0.
2. Выбор следующей записи вполне логично для SQLSelect, а не для всех типов SQL.
3. DoAfterExecute как следует из названия должен выполнялся самой последней инструкцией, после всех действий.

Лучше явно управлять транзакциями, чем пытаясь вылечить "автоматический" режим "сломать" основной функционал.

**Comment by madorin** [comment]
Привет, AnryJester,
Да, щас проблема в том что DoAfterExecute вызывается до Next и собственно получается RecordCount = 0.
Хотя незнаю если правильнее так как было. Будем править, чтобы не ломать совместимость.

**Comment by madorin** [comment]
Anry, смотрел код.

В 7.6 DoAfterExecute и выполняется до Next, независимо от этого фикса. этого было до.
В 7.5 было иначе да, сначала Next а потом DoAfterExecute.
Я нахожу логику в 7.6 более правильной.
В DoAfterExecute RecordCound максимум может быть 1, после одного фетча что позволяет проверять только если есть записи и ничего другое.
Посмотрим что скажет сообщество. Можно фиксить только ради совместимости.

> 2. Выбор следующей записи вполне логично для SQLSelect, а не для всех типов SQL.

Еще для сингулярных запросов тоже, типа execute procedure. Хотя в FB3 уже это не работает. Уже курсор с одной записи не открывается.

> 3. DoAfterExecute как следует из названия должен выполнялся самой последней инструкцией, после всех действий.

Next уже как бы не связан с Execute. есть isc_execute а есть isc_fetch.

**Comment by vkrapotkin** [comment]
Я бы вообще описал RecordCount как deprecated
По идее, нужно бы пользоваться явным RecordCountFromSrv
